File: app/src/main/java/workshop1024/com/xproject/controller/activity/PublisherActivity.java
package workshop1024.com.xproject.controller.activity;

import android.content.res.Resources;
import android.os.Bundle;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.app.DialogFragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.Toast;

import java.util.List;

import workshop1024.com.xproject.R;
import workshop1024.com.xproject.controller.adapter.PublisherListAdapter;
import workshop1024.com.xproject.model.publisher.Publisher;
import workshop1024.com.xproject.model.publisher.source.PublisherDataSource;
import workshop1024.com.xproject.model.publisher.source.PublisherRepository;
import workshop1024.com.xproject.model.publisher.source.local.PublisherDatabase;
import workshop1024.com.xproject.model.publisher.source.local.PublisherLocalDataSource;
import workshop1024.com.xproject.model.publisher.source.remote.PublisherRemoteDataSource;
import workshop1024.com.xproject.utils.ExecutorUtils;
import workshop1024.com.xproject.view.RecyclerViewItemDecoration;
import workshop1024.com.xproject.view.dialog.SingleChoiceDialog;

/**
 * 发布者列表页面
 */
public class PublisherActivity extends AppCompatActivity implements SwipeRefreshLayout.OnRefreshListener,
        PublisherDataSource.LoadPublishersCallback, SingleChoiceDialog.SingleChoiceDialogListener,
        PublisherListAdapter.OnPublisherListSelectListener {

    private CoordinatorLayout mRootView;
    private Toolbar mToolbar;
    private SwipeRefreshLayout mPublisherSwipeRefreshLayout;
    private RecyclerView mPublisherRecyclerView;

    private SingleChoiceDialog mTypeChoiceDialog;
    private SingleChoiceDialog mLanguageChoiceDialog;
    private DialogFragment mSelectedDialog;

    private PublisherListAdapter mPublisherListAdapter;
    private PublisherRepository mPublisherRepository;

    //可选择的发布者
    private String[] mSelectTypeStrings;
    //可选择的语言
    private String[] mSelectLanguageStrings;
    //选择的发布者索引
    private int mSelectedTypeIndex = 0;
    //选择的语言索引
    private int mSelectedLanguageIndex;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_publisher);

        mRootView = findViewById(R.id.publisher_coordinatorlayout_rootview);
        mToolbar = findViewById(R.id.publisher_toolbar_navigator);
        mPublisherSwipeRefreshLayout = findViewById(R.id.publisher_swiperefreshlayout_pullrefresh);
        mPublisherRecyclerView = findViewById(R.id.publisher_recyclerview_list);

        setSupportActionBar(mToolbar);
        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);

        mPublisherSwipeRefreshLayout.setOnRefreshListener(this);
        mPublisherRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        mPublisherRecyclerView.addItemDecoration(new RecyclerViewItemDecoration(6));

        //获取选择的发布者和语言字符串
        Resources resources = getResources();
        mSelectTypeStrings = resources.getStringArray(R.array.publisher_dialog_publishers);
        mSelectLanguageStrings = resources.getStringArray(R.array.language_dialog_languages);

        //显示默认选中的发布者
        //mToolbar.setTitle()在此处不生效，参考：https://stackoverflow
        // .com/questions/26486730/in-android-app-toolbar-settitle-method-has-no-effect-application-name-is-shown
        actionBar.setTitle(mSelectTypeStrings[mSelectedTypeIndex]);

        //使用默认选中的发布者类型请求发布者信息
        mPublisherRepository = PublisherRepository.getInstance(PublisherRemoteDataSource.getInstance(),
                PublisherLocalDataSource.getInstance(PublisherDatabase.getInstance(this).publisherDao(), new
                        ExecutorUtils()));
//        mPublisherRepository.refreshLimitedPublishers();
        mPublisherSwipeRefreshLayout.setRefreshing(true);
        mPublisherRepository.getPublishersByType(mSelectTypeStrings[mSelectedTypeIndex], this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater menuInflater = getMenuInflater();
        menuInflater.inflate(R.menu.publisher_toolbar_menu, menu);

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.publisher_menu_search:
                Toast.makeText(this, "publisher_menu_search", Toast.LENGTH_SHORT).show();
                break;
            case R.id.publisher_menu_filter:
                //FIXME 每次都要创建一个对象吗？
                mTypeChoiceDialog = SingleChoiceDialog.newInstance(R.string.publisher_dialog_title,
                        mSelectTypeStrings, mSelectedTypeIndex);
                mTypeChoiceDialog.show(getSupportFragmentManager(), "ChoiceTypeDialog");
                break;
            case R.id.publisher_menu_language:
                mLanguageChoiceDialog = SingleChoiceDialog.newInstance(R.string.language_dialog_title,
                        mSelectLanguageStrings, mSelectedLanguageIndex);
                mLanguageChoiceDialog.show(getSupportFragmentManager(), "ChoiceLanguageDialog");
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onRefresh() {
        mPublisherSwipeRefreshLayout.setRefreshing(true);
        mPublisherRepository.refreshPublishers(mSelectedDialog.getTag());
        if (mSelectedDialog == mTypeChoiceDialog) {
            mPublisherRepository.getPublishersByType(mSelectTypeStrings[mSelectedTypeIndex], this);
        } else if (mSelectedDialog == mLanguageChoiceDialog) {
            mPublisherRepository.getPublishersByLanguage(mSelectLanguageStrings[mSelectedLanguageIndex], this);
        }
    }

    @Override
    public void onSingleChoiceDialogItemClick(DialogFragment dialog, String selectItem) {
        mSelectedDialog = dialog;
        mToolbar.setTitle(selectItem);

        mPublisherSwipeRefreshLayout.setRefreshing(true);

        if (dialog == mTypeChoiceDialog) {
            mPublisherRepository.getPublishersByType(selectItem, this);
        } else if (dialog == mLanguageChoiceDialog) {
            mPublisherRepository.getPublishersByLanguage(selectItem, this);
        }
    }

    @Override
    public void publisherListItemSelect(Publisher selectPublisher, boolean isSelected) {
        if (isSelected) {
            Snackbar.make(mRootView, selectPublisher.getName() + " selected", Snackbar.LENGTH_SHORT).show();
            mPublisherRepository.subscribePublisherById(selectPublisher.getPublisherId());
        } else {
            Snackbar.make(mRootView, selectPublisher.getName() + " unselected", Snackbar.LENGTH_SHORT).show();
            mPublisherRepository.unSubscribePublisherById(selectPublisher.getPublisherId());
        }
    }

    @Override
    public void onPublishersLoaded(List<Publisher> publisherList) {
        //FIXME 每次都需要创建适配器吗？
        mPublisherSwipeRefreshLayout.setRefreshing(false);
        mPublisherListAdapter = new PublisherListAdapter(publisherList, this);
        mPublisherRecyclerView.setAdapter(mPublisherListAdapter);
    }

    @Override
    public void onDataNotAvailable() {

    }
}

File: main/src/main/java/workshop1024/com/xproject/main/controller/activity/PublisherActivity.kt
package workshop1024.com.xproject.main.controller.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.annotation.VisibleForTesting
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.DialogFragment
import com.google.android.material.snackbar.Snackbar
import workshop1024.com.xproject.base.controller.activity.XActivity
import workshop1024.com.xproject.base.view.recyclerview.RecyclerViewItemDecoration
import workshop1024.com.xproject.main.R
import workshop1024.com.xproject.main.controller.adapter.PublisherListAdapter
import workshop1024.com.xproject.main.controller.idlingResource.PublisherIdlingResouce
import workshop1024.com.xproject.main.databinding.PublisherActivityBinding
import workshop1024.com.xproject.main.model.Injection
import workshop1024.com.xproject.main.model.publisher.Publisher
import workshop1024.com.xproject.main.model.publisher.PublisherType
import workshop1024.com.xproject.main.model.publisher.source.PublisherDataSource
import workshop1024.com.xproject.main.view.dialog.TypeChoiceDialog

/**
 * 发布者列表页面
 */
class PublisherActivity : XActivity(), PublisherDataSource.LoadPublisherAndPublisherTypeCallback, TypeChoiceDialog.TypeChoiceDialogListener, PublisherListAdapter.OnPublisherListSelectListener {

    private var mTypeChoiceDialog: TypeChoiceDialog? = null
    private var mLanguageChoiceDialog: TypeChoiceDialog? = null
    private var mSelectedDialog: DialogFragment? = null

    private var mPublisherListAdapter: PublisherListAdapter? = null

    private lateinit var mPublisherRepository: PublisherDataSource

    //可选择发布者内容类型
    private var mContentTypeList: ArrayList<PublisherType>? = null
    //可选择发布者语言类型
    private var mLanguageTypeList: ArrayList<PublisherType>? = null
    //选择的发布者索引
    private val mSelectedTypeIndex = 0
    //选择的语言索引
    private val mSelectedLanguageIndex = 0

    private lateinit var mPublisherActivityBinding: PublisherActivityBinding

    //发布者dlingResouce，检测请求发布者列表异步任务
    @VisibleForTesting
    var mPublisherIdlingResouce: PublisherIdlingResouce? = null
        get() {
            if (field == null) {
                field = PublisherIdlingResouce()
            }
            return field
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mPublisherActivityBinding = DataBindingUtil.setContentView(this, R.layout.publisher_activity)

        setSupportActionBar(mPublisherActivityBinding.publisherToolbarNavigator)
        val actionBar = supportActionBar
        actionBar!!.setDisplayHomeAsUpEnabled(true)

        mPublisherActivityBinding.publisherSwiperefreshlayoutPullrefresh.isEnabled = false
        mPublisherActivityBinding.publisherRecyclerviewList.addItemDecoration(RecyclerViewItemDecoration(6))

        //显示默认选中的发布者
        //mToolbar.setTitle()在此处不生效，参考：https://stackoverflow
        // .com/questions/26486730/in-android-app-toolbar-settitle-method-has-no-effect-application-mName-is-shown
        actionBar.title = resources.getString(R.string.publisher_title_default)

        mPublisherRepository = Injection.providePublisherRepository(this)
    }

    override fun onStart() {
        super.onStart()
        mPublisherActivityBinding.publisherSwiperefreshlayoutPullrefresh.isRefreshing = true
        //开始请求发布者列表，IdlingResouce设置为false，等待异步执行完毕
        mPublisherIdlingResouce?.setIdleState(false)
        mPublisherRepository.getPublishersAndPublisherTypes(this)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.publisher_toolbar_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.publisher_menu_search -> Toast.makeText(this, "publisher_menu_search", Toast.LENGTH_SHORT).show()
            R.id.publisher_menu_filter -> {
                //FIXME 每次都要创建一个对象吗？
                mTypeChoiceDialog = TypeChoiceDialog.newInstance(R.string.publisher_dialog_title,
                        mContentTypeList!!, mSelectedTypeIndex)
                mTypeChoiceDialog!!.show(supportFragmentManager, "ChoiceTypeDialog")
            }
            R.id.publisher_menu_language -> {
                mLanguageChoiceDialog = TypeChoiceDialog.newInstance(R.string.language_dialog_title,
                        mLanguageTypeList!!, mSelectedLanguageIndex)
                mLanguageChoiceDialog!!.show(supportFragmentManager, "ChoiceLanguageDialog")
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onCacheOrLocalPublishersLoaded(publisherList: List<Publisher>) {
        //FIXME 每次都需要创建适配器吗？
        refreshPublisherList(publisherList)
        Snackbar.make(mPublisherActivityBinding.root, "Fetch cacheorlocal " + publisherList.size + " publishers ...", Snackbar.LENGTH_SHORT).show()
        if (!mPublisherRepository.getIsRequestRemote()) {
            mPublisherActivityBinding.publisherSwiperefreshlayoutPullrefresh.isRefreshing = false
        }
        mPublisherIdlingResouce?.setIdleState(true)
    }

    override fun onRemotePublishersLoaded(publisherList: List<Publisher>) {
        refreshPublisherList(publisherList)
        Snackbar.make(mPublisherActivityBinding.root, "Fetch remote " + publisherList.size + " publishers ...", Snackbar.LENGTH_SHORT).show()
        mPublisherActivityBinding.publisherSwiperefreshlayoutPullrefresh.isRefreshing = false
        //请求发布者类表完毕，IdlingResouce设置为true，执行后面异步指令
        mPublisherIdlingResouce?.setIdleState(true)
    }


    override fun onCacheOrLocalPublisherTypesLoaded(contentTypeList: List<PublisherType>, languageTyleList: List<PublisherType>) {
        refreshPublisherTypeList(contentTypeList, languageTyleList)
    }

    override fun onRemotePublisherTypesLoaded(contentTypeList: List<PublisherType>, languageTyleList: List<PublisherType>) {
        refreshPublisherTypeList(contentTypeList, languageTyleList)
    }

    private fun refreshPublisherTypeList(contentTypeList: List<PublisherType>, languageTyleList: List<PublisherType>) {
        mContentTypeList = contentTypeList as ArrayList<PublisherType>
        mLanguageTypeList = languageTyleList as ArrayList<PublisherType>
    }

    private fun refreshPublisherList(publisherList: List<Publisher>) {
        mPublisherListAdapter = PublisherListAdapter(publisherList, this)
        mPublisherActivityBinding.publisherRecyclerviewList.adapter = mPublisherListAdapter
    }

    override fun onDataNotAvailable() {

    }

    override fun onTypeChoiceDialogItemClick(dialog: DialogFragment, publisherType: PublisherType) {
        mSelectedDialog = dialog
        mPublisherActivityBinding.publisherToolbarNavigator.title = publisherType.mName

        mPublisherActivityBinding.publisherSwiperefreshlayoutPullrefresh.isRefreshing = true
        if (dialog === mTypeChoiceDialog) {
            mPublisherRepository.getPublishersByContentType(publisherType.mTypeId, this)
        } else if (dialog === mLanguageChoiceDialog) {
            mPublisherRepository.getPublishersByLanguageType(publisherType.mTypeId, this)
        }
    }

    override fun onPublisherListItemSelect(selectPublisher: Publisher, isSelected: Boolean) {
        if (isSelected) {
            //FIXME 订阅网络请求返回后，在提示并且选中
            mPublisherRepository.subscribePublisherById(selectPublisher.mPublisherId)
            Snackbar.make(mPublisherActivityBinding.root, selectPublisher.mName + " selected", Snackbar.LENGTH_SHORT).show()
        } else {
            mPublisherRepository.unSubscribePublisherById(selectPublisher.mPublisherId)
            Snackbar.make(mPublisherActivityBinding.root, selectPublisher.mName + " unselected", Snackbar.LENGTH_SHORT).show()
        }
    }

    companion object {

        fun startActivity(context: Context) {
            val intent = Intent(context, PublisherActivity::class.java)
            context.startActivity(intent)
        }
    }
}